runtime/Java/src/org/antlr/v4/runtime/atn/ParserATNSimulator.java

```diff
@@ -1297,65 +1297,51 @@ else if (altToPred[i] != SemanticContext.NONE) {
 	protected int getSynValidOrSemInvalidAltThatFinishedDecisionEntryRule(ATNConfigSet configs,
 																		  ParserRuleContext outerContext)
 	{
-		Pair<ATNConfigSet,ATNConfigSet> sets =
-			splitAccordingToSemanticValidity(configs, outerContext);
-		ATNConfigSet semValidConfigs = sets.a;
-		ATNConfigSet semInvalidConfigs = sets.b;
-		int alt = getAltThatFinishedDecisionEntryRule(semValidConfigs);
-		if ( alt!=ATN.INVALID_ALT_NUMBER ) { // semantically/syntactically viable path exists
-			return alt;
-		}
-		// Is there a syntactically valid path with a failed pred?
-		if ( semInvalidConfigs.size()>0 ) {
-			alt = getAltThatFinishedDecisionEntryRule(semInvalidConfigs);
-			if ( alt!=ATN.INVALID_ALT_NUMBER ) { // syntactically viable path exists
-				return alt;
+		// The minimum alternative that finished the decision entry rule and is
+		// valid according to semantic predicates.
+		int minAltSucceeded = Integer.MAX_VALUE;
+		// The minimum alternative that finished the decision entry rule but is
+		// not valid according to semantic predicates. Only considered if
+		// minAltSucceeded does not produce an alternative.
+		int minAltFailed = Integer.MAX_VALUE;
+		for (ATNConfig c : configs) {
+			if (c.alt >= minAltSucceeded) {
+				// This alternative is no better than our current best
+				continue;
 			}
-		}
-		return ATN.INVALID_ALT_NUMBER;
-	}
 
-	protected int getAltThatFinishedDecisionEntryRule(ATNConfigSet configs) {
-		IntervalSet alts = new IntervalSet();
-		for (ATNConfig c : configs) {
-			if ( c.getOuterContextDepth()>0 || (c.state instanceof RuleStopState && c.context.hasEmptyPath()) ) {
-				alts.add(c.alt);
+			if (c.getOuterContextDepth() == 0 && !(c.state instanceof RuleStopState && c.context.hasEmptyPath())) {
```

Comment:
📝 This condition should be verified in detail next to the original, which was part of `getAltThatFinishedDecisionEntryRule`